source/hpack/tables.d

```diff
@@ -0,0 +1,340 @@
+	@property size_t size() @safe { return m_size; }
+
+	@property size_t index() @safe { return m_index; }
+
+	@property ref auto table() @safe { return m_table; }
+
+	HTTP2HeaderTableField opIndex(size_t idx) @safe
+	{
+		foreach(i,f; m_table[].enumerate(1)) {
+			 if(i == idx) return f;
+		}
+		assert(false, "Invalid dynamic table index");
+	}
+
+	// insert at the head
+	void insert(HTTP2HeaderTableField header) @safe
+	{
+		auto nsize = computeESize(header);
+		// ensure that the new entry does not exceed table capacity
+		while(m_size + nsize > m_maxsize) {
+			logDebug("Maximum header table size exceeded");
+			remove();
+		}
+
+		// insert
+		m_table.insertFront(header);
+		m_size += nsize;
+		m_index++;
+	}
+
+	// evict an entry
+	void remove() @safe
+	{
+		assert(!m_table.empty, "Cannot remove element from empty table");
+		m_size -= computeESize(m_table.back);
+		m_table.removeBack();
+		m_index--;
+	}
+
+	/** new size should be lower than the max set one
+	  * after size is successfully changed, an ACK has to be sent
+	  * multiple changes between two header fields are possible
+	  * if multiple changes occour, only the smallest maximum size
+	  * requested has to be acknowledged
+	*/
+	void updateMaxSize(size_t nsize) @safe
+	{
+		assert(false);
+	}
+
+	// compute size of an entry as per RFC
+	private size_t computeESize(HTTP2HeaderTableField f) @safe
+	{
+		return f.name.sizeof + f.value.sizeof + 32;
+	}
+}
+
+unittest {
+	// static table
+	auto a = StaticTable[1];
+	static assert(is(typeof(a) == immutable(HTTP2HeaderTableField)));
+	assert(a.name == ":authority");
+	assert(StaticTable[2].name == ":method" && StaticTable[2].value == HTTPMethod.GET);
+
+	HTTP2Settings settings;
+
+	DynamicTable dt = DynamicTable(settings.headerTableSize);
+	assert(dt.size == 0);
+	assert(dt.index == 0);
+
+	// dynamic table
+	import std.algorithm.comparison : equal;
+
+	auto h = HTTP2HeaderTableField("test", "testval");
+	dt.insert(h);
+	assert(dt.size > 0);
+	assert(dt.index == 1);
+	assert(equal(dt.table[], [h]));
+	assert(dt.table[].front.name == "test");
+	assert(dt[dt.index].name == "test");
+
+	dt.remove();
+	assert(dt.size == 0);
+	assert(dt.index == 0);
+}
+
+/** provides an unified address space through operator overloading
+  * this is the only interface that will be used for the two tables
+  */
+struct IndexingTable {
+	private {
+		alias StaticTable m_static;
+		DynamicTable m_dynamic;
+	}
+
+	// requires the maximum size for the dynamic table
+	this(HTTP2SettingValue ms) @safe
+	{
+		m_dynamic = DynamicTable(ms);
+	}
+
+	@property size_t size() @safe { return STATIC_TABLE_SIZE + m_dynamic.index; }
+
+	// element retrieval
+	HTTP2HeaderTableField opIndex(size_t idx) @safe
+	{
+		assert(idx <= size(), "Invalid table index");
+
+		if (idx <= STATIC_TABLE_SIZE) return m_static[idx];
+		else return m_dynamic[idx-STATIC_TABLE_SIZE];
+	}
+
+	// forward to insert
+	auto opOpAssign(string op)(HTTP2HeaderTableField hf) @safe
```

Comment:
Using arithmetic operators for non-numeric operations is quite non-idiomatic style in D, as opposed to C++, and I'd probably just stick to the `insert()` method to avoid confusion at the call site.